### ISSUE TITLE
Make searchsorted and bucketize API consistent

### ISSUE THREAD

**Issue by hassec** [pytorch/pytorch#54588]
## 🐛 Bug

I was a bit surprised by the fact that the API of `searchsorted` and `bucketize` is different, since the first two arguments are in opposite order.

I was originally using `searchsorted` with a 1-D array and noticed the following *Note* in the documentation:
> If your use case is always 1-D sorted sequence, `torch.bucketize()` is preferred, because it has fewer dimension checks resulting in slightly better performance.

So I switched to use `bucketize` and noticed that the API is incompatible, which I think is an unnecessary inconvenience from the user perspective. (Of course I'm not sure if this was a conscious decision based on information I'm not aware of :P )
But if not, is it possible to align the two?  

Currently: 
```py
torch.searchsorted(sorted_sequence, values, ...)
torch.bucketize(input, boundaries, ...)
```
I think more natural would be to align `bucketize` to `searchsorted` by swapping the arguments and IMHO also renaming them as I find the `searchsorted` naming more self-documenting.

Small additional sidenote re the *Note* in the documentation. I wasn't really able to see any speed difference between the two.
When looking at [Bucketization.cu](https://github.com/pytorch/pytorch/blob/2ecb2c79312eac6924182fe343dc3cd9a205305c/aten/src/ATen/native/cuda/Bucketization.cu) and [Bucketization.cpp](https://github.com/pytorch/pytorch/blob/2ecb2c79312eac6924182fe343dc3cd9a205305c/aten/src/ATen/native/Bucketization.cpp)  it seems that `bucketize` really just forwards its search to `searchsorted` so the only speedup could potentially come, from the additional `__builtin_expect` which is injected via the  `TORCH_CHECK` call in `bucketize`. But was it ever actually checked that this enabled the compiler to do something smarter and thus make `bucketize` faster ?

**Comment by mruberry** [comment]
That is frustrating!

[torch.bucketize](https://pytorch.org/docs/master/generated/torch.bucketize.html?highlight=bucketize#torch.bucketize) is supposed to be based on TF's [bucketize](https://www.tensorflow.org/tfx/transform/api_docs/python/tft/bucketize):

https://github.com/pytorch/pytorch/blob/67d44377e34a4194f9899391097ec52658a44af4/aten/src/ATen/native/Bucketization.cpp#L5

but they don't seem similar to me. torch.bucketize also appears to be very little-used. We should probably deprecate it.

[torch.searchsorted](https://pytorch.org/docs/master/generated/torch.searchsorted.html?highlight=searchsorted#torch.searchsorted), on the other hand, should be compatible with NumPy's [searchsorted](https://numpy.org/doc/stable/reference/generated/numpy.searchsorted.html?highlight=searchsorted#numpy.searchsorted), but it looks like it needs a refresh, too.

I've added some labels and made a note of the discrepancies. We would accept PRs that make torch.searchsorted more NumPy compatible by adding kwargs corresponding to NumPy's kwargs, for example, and we would accept a PR that removes the suggestion that bucketize is to be preferred.

**Comment by hassec** [comment]
I guess removing the note from the documentation should be easy. I'm happy to do that.

But I'm quite sure what would the best way is to align torch's `searchsorted` more with that of numpy.    
Numpy, allows us to pass a "sorter", which torch currently doesn't. 
For the 1-D case which numpy restricts itself to, this is usually just the result of `argsort` which can then easily be used. 
For the N-D case this, while still possible, IMHO seems like unnecessary complications.

**Comment by mruberry** [comment]
> Thanks for the feedback @mruberry 😊

You're welcome ;)

> I guess removing the note from the documentation should be easy. I'm happy to do that.

Thank you!

> But I'm quite sure what would the best way is to align torch's `searchsorted` more with that of numpy.
> Numpy, allows us to pass a "sorter", which torch currently doesn't.
> For the 1-D case which numpy restricts itself to, this is usually just the result of `argsort` which can then easily be used.
> For the N-D case this, while still possible, IMHO seems like unnecessary complications.

I would also need to think about it more, but at least this issue will let us track the challenge.

**Comment by hassec** [comment]
@mruberry sorry I tried.   
The `CONTRIBUTING.md` didn't mention I need to sign a CLA.... 

I do **not** want to start any discussion on the topic, but to be 100% honest, just don't have the time/interest right now to figure out if I want to sign a CLA.    

Thanks again for your feedback, and sorry I can't be of more help.

**Comment by mruberry** [comment]
> @mruberry sorry I tried.
> The `CONTRIBUTING.md` didn't mention I need to sign a CLA....
> 
> I do **not** want to start any discussion on the topic, but to be 100% honest, just don't have the time/interest right now to figure out if I want to sign a CLA.
> 
> Thanks again for your feedback, and sorry I can't be of more help.

I understand, thanks @hassec.
